src/frontend/Account.ts

```diff
@@ -42,79 +43,61 @@ class Account {
    * login with with magic, get provider, signer and set up
    * the smart contracts.
    * @param email user's email address
-   * @param walletType The type of wallet (metamask or magic)
-   * @param network the name of the (polygon) network.
    */
-  async connect(args: IConnectAccount): Promise<void> {
-    if (this.isConnected(args.walletType)) return;
-
-    await this.changeWallet(args);
-
-    console.time("init-contracts");
-    await init(this.signer, this.biconomy);
-    console.timeEnd("init-contracts");
-  }
-
-  async setSimulationMode(mode: boolean) {
-    return setSimulationMode(mode);
-  }
-
-  async changeWallet({
+  async connect({
     walletType,
     email,
+    shouldRunMagicTestMode,
     getMessage,
     verify,
-    shouldRunMagicTestMode,
-  }: IConnectAccount): Promise<void | undefined> {
+    chainId,
+  }: IConnectAccount): Promise<void> {
+    console.log("connected: ", this.isConnected(walletType, chainId));
+    if (this.isConnected(walletType, chainId)) return;
+
+    // get wallet provider based on wallet type
     let walletProvider: ethers.providers.Web3Provider | undefined;
-    // change to metamask
     if (walletType === "magic" && email) {
-      const { magic, provider } = await initMagic({ email, testMode: Boolean(shouldRunMagicTestMode) });
+      const { magic, provider } = await initMagic({ email, testMode: Boolean(shouldRunMagicTestMode), chainId });
 
       if (magic) this.magic = magic;
       walletProvider = provider;
-    } else if (walletType === "metamask" || walletType === "coinbase") {
-      if (walletType === "metamask" && !window.ethereum) return;
-      // we know that window.ethereum exists here
-      const _provider = new ethers.providers.Web3Provider(
-        (await getExternalProvider(walletType)) as ethers.providers.ExternalProvider,
-      );
-
-      // requires requesting permission to connect users accounts
-      await _provider.send("eth_requestAccounts", []);
-
-      walletProvider = _provider;
-    } else if (walletType === "walletConnect") {
-      // walletConnect doesn't require window.ethereum to be present always
-      const provider = (await getExternalProvider(walletType)) as WalletConnectProvider;
-
-      if (provider) this.walletConnectProvider = provider;
-
-      if (provider) {
-        walletProvider = new ethers.providers.Web3Provider(provider as ethers.providers.ExternalProvider);
-      }
+    } else {
+      walletProvider = await getWeb3Provider(walletType, chainId);
```

Comment:
Why do we need to pass the chainId in order to get the provider?

Don't all the wallets (besides Magic) have something like `window.ethereum` that will provide the chainId for you?

I see [here](https://github.com/AffineLabs/defi-sdk/blob/0b22a1ca5f5b9e50a1c54d360b543097ae516ee8/src/frontend/wallets.ts#L70-L91) that for coinbase and wallet connect you're taking our alchemy url and using that as a provider. Is that necessary?

Reply:
No, both WalletConnect & Magic don't inject providers (like `window.ethereum`) into the browser. That's why we need to provide our RPC URLs to this kind of wallet.